Add the Antarctic mesh creation within the landice core

hollyhan opened the pull request:
Add a Antarctic mesh creation testcase following the addition of the Humbolt case in PR #322.

xylar commented on compass/landice/__init__.py:
```diff
@@ -28,3 +29,4 @@ def __init__(self):
         self.add_test_group(Humboldt(mpas_core=self))
         self.add_test_group(HydroRadial(mpas_core=self))
         self.add_test_group(Thwaites(mpas_core=self))
+        self.add_test_group(Antarctic(mpas_core=self))
```
We like to keep these in alphabetical order. Could you move this to the top?

xylar commented on compass/landice/tests/antarctic/__init__.py:
```diff
@@ -0,0 +1,17 @@
+
+class Antarctic(TestGroup):
+    """
+    A test group for antarctic test cases
```
```suggestion
    A test group for Antarctic test cases
```

xylar commented on compass/landice/tests/antarctic/default/__init__.py:
```diff
@@ -0,0 +1,28 @@
+
+class Default(TestCase):
+    """
+    The default test case for the humboldt test group simply creates the
```
```suggestion
    The default test case for the Antarctic test group simply creates the
```

xylar commented on compass/landice/tests/antarctic/mesh.py:
```diff
@@ -0,0 +1,206 @@
+
+class Mesh(Step):
+    """
+    A step for creating a mesh and initial condition for antarctica test cases
```
```suggestion
    A step for creating a mesh and initial condition for Antarctic test cases
```

xylar commented on compass/landice/tests/antarctic/mesh.py:
```diff
@@ -0,0 +1,206 @@
+        self.add_output_file(filename='graph.info')
+        self.add_output_file(filename='Antarctic_1to10km.nc')
+        self.add_input_file(
+                filename='antarctica_8km_2020_10_20.nc',
```
Just a question for everyone, is the 8 km data set good enough for this purpose or would you want to use something down to 1 km since the target mesh is higher res?  I assume this is BedMachine topography, which I seem to recall is available at 500 m resolution.

trhille replied:
We will very likely end up using a 1km dataset eventually, but this makes testing much faster.

xylar reviewed (commented):
I know this is labelled as a "in progress".  I just wanted to make a few small comments that should be easy to address and then ask a bigger question about the resolution of the input data set.

xylar commented on compass/landice/tests/antarctic/mesh.py:
```diff
@@ -0,0 +1,206 @@
+        config = self.config
+        section = config['antarctic']
+
+        logger.info('calling build_cell_wdith')
```
```suggestion
        logger.info('calling build_cell_width')
```

xylar commented:
When testing this, I'm seeing the issue I created in MPAS-Tools: https://github.com/MPAS-Dev/MPAS-Tools/issues/454

xylar commented on compass/landice/tests/antarctic/mesh.py:
```diff
@@ -0,0 +1,206 @@
+        super().__init__(test_case=test_case, name='mesh')
+
+        self.add_output_file(filename='graph.info')
+        self.add_output_file(filename='Antarctic_1to10km.nc')
```
It's a little tricky to do but maybe we want the name of the output mesh to include the actual requested minimum and maximum resolution, rather than these hard-coded values?  Something to discuss.

trhille replied:
I think something like this would work, but is ugly to look at:
`      config = self.config
        section = config['humboldt']
        min_spac = float(section.get('min_spac'))
        max_spac = float(section.get('max_spac'))
        min_spac_str = f"{round(min_spac/1000, 2):.1f}"
        max_spac_str = f"{round(max_spac/1000, 2):.1f}"
        if min_spac_str[-1] == '0':
             min_spac_str_trimmed =  min_spac_str[0:-2]
        else:
             min_spac_str_trimmed = min_spac_str

        if max_spac_str[-1] == '0':
             max_spac_str_trimmed =  max_spac_str[0:-2]
        else:
             max_spac_str_trimmed = max_spac_str

        output_file_name = 'Antarctic_{}to{}km.nc'.format(min_spac_str_trimmed, max_spac_str_trimmed))` 
                       
We could add a function to `compass.landice.mesh` that does this kind of filename formatting.

xylar replied:
Yes, that's roughly what I was thinking of.  But a serious complication is that the file name isn't known until the `run()` method because a user could alter the config options before then.  That makes it pretty hard to define the output name either in the `__init__()` method or the `setup()` method.  Maybe in the end it would be better not to have the resolution in the file name at all?  Or maybe we're fine with hard coding the resolution even if a user might change it.  I just want to have the discussion so this doesn't happen without thinking it through.

trhille replied:
Okay, I see what you mean. Our convention for MALI mesh names includes the resolution, but also other things like the date and a version number because of all the iterations we end up doing with the optimization and initial condition datasets, so I'd be totally fine with not including resolution in the output file name within COMPASS. @matthewhoffman, what do you think?

matthewhoffman replied:
I agree we could leave it out given the versioning system we have been (mostly) using successfully.  But this is a valid concern that we may want to revisit down the load if/when our mesh generation workflow is truly end to end.

xylar commented on compass/landice/tests/antarctic/antarctic.cfg:
```diff
@@ -0,0 +1,29 @@
+cull_distance = 5.0
+
+# mesh density parameters
+# minimum cell spacing (meters)
+min_spac = 1.e4
+# maximum cell spacing (meters)
+max_spac = 1.e5
```
These are presumably for testing but they're 10 times coarser than the mesh name suggests.

hollyhan replied:
Thanks for pointing it out, @xylar. That's right that it these coarse min/max cell spacing were defined for a quick testing, and I didn't feel the need to change the output filename that is hard-coded in the main code `mesh.py`. However, I've switched the values back to `min_spac = 1.e3` and `max_spac = 1.e4` in this file to be consistent within this PR still.

xylar replied:
I still think it's worth discussing the file name a bit further. If I were a user and I changed the resolution, I would be mildly annoyed that the file name wasn't right.

xylar commented on compass/landice/tests/antarctic/antarctic.cfg:
```diff
@@ -0,0 +1,29 @@
+high_log_speed = 2.5
+# log10 of min speed (m/yr) for cell spacing
+low_log_speed = 0.75
+# distance at which cell spacing = max_spac (meters)
+high_dist = 1.e5
+# distance within which cell spacing = min_spac (meters)
+low_dist = 1.e4
```
Unless I missed something, these aren't being used (yet?).

trhille replied:
Those get used here: https://github.com/hollyhan/compass/blob/compass_add_AISmeshGen/compass/landice/mesh.py#L177-L179

xylar replied:
Thanks @trhille. In that case, the comments here should clarify what the distance is to or between.

xylar replied:
Also, the shared function should probably document, rather than just list, these config options. We don’t need to change that in this PR, though.

trhille replied:
@hollyhan, Yeah that's correct. But maybe say `distance to ice edge or grounding line`. Currently, we use `high_dist` and `low_dist` in both the `use_distance_to_grounding_line` and  `use_distance_to_ice_edge` cases, which should probably be changed eventually.

pep8speaks commented:
Hello @hollyhan! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`compass/landice/mesh.py`](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/mesh.py):

> [Line 178:70](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/mesh.py#L178): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 179:22](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/mesh.py#L179): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 179:73](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/mesh.py#L179): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets





* In the file [`compass/landice/tests/antarctic/mesh.py`](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/tests/antarctic/mesh.py):

> [Line 62:52](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/tests/antarctic/mesh.py#L62): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 83:80](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/tests/antarctic/mesh.py#L83): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 79 characters)
> [Line 89:1](https://github.com/MPAS-Dev/compass/blob/2aaa20cea48011ee8238281bbae7eb258a121f3d/compass/landice/tests/antarctic/mesh.py#L89): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

matthewhoffman commented:
@hollyhan , as we discussed today, go ahead and fix the PEP8 issues on this PR, and then I'll review it and we can try to get it merged quickly.

hollyhan commented:
Thanks for the reivew, @matthewhoffman.
I've addressed the PEP 8 issues and updated the PR.

xylar commented on compass/landice/mesh.py:
```diff
@@ -201,7 +214,7 @@ def set_cell_width(self, section, thk, vx=None, vy=None,
     # that distance.
     if dist_to_edge is not None:
         cell_width[np.logical_and(thk == 0.0,
-                   dist_to_edge > (10. * cull_distance))] = max_spac
+                                  dist_to_edge > (10. * cull_distance))] = max_spac
```
I'd suggest:
``` python
    if dist_to_edge is not None:
        mask = np.logical_and(
            thk == 0.0, dist_to_edge > (10. * cull_distance))
        cell_width[mask] = max_spac
```

xylar commented:
@hollyhan, it looks like this won't be able to be merged until you rebase (because another merge also modified `compass/landice/mesh.py`).  We can work on that tomorrow if you like.

xylar commented on compass/landice/mesh.py:
```diff
@@ -304,8 +306,8 @@ def get_dist_to_edge_and_GL(self, thk, topg, x, y, window_size=1.e5):
 
-        dist_to_here = ((XPOS[np.ix_(irng, jrng)] - x[j])**2 +
-                        (YPOS[np.ix_(irng, jrng)] - y[i])**2)**0.5
+        dist_to_here = ((XPOS[np.ix_(irng, jrng)] - x[j]) ** 2 +
+                        (YPOS[np.ix_(irng, jrng)] - y[i]) ** 2) ** 0.5
```
Out of curiosity, is this one that PEP8 complained about?  I would never put spaces around a `**` operator myself.  I don't think it's very intuitive.  So it would be annoying to me to find out that I'm breaking the rules.

hollyhan replied:
@xylar  PEP8 didn't complain about this. There is another line that puts spaces around `**` (`speed = (vx ** 2 + vy ** 2) ** 0.5`) and multiple lines that put spaces round `* and other operaters `-`, `/`, etc. I can change them if you prefer. And if we decide to change them, we would also need to change the same `mesh.py` files in the Greenland-mesh-generation test case.

xylar replied:
Feel free to leave it. I agree, the important thing is to be consistent. I like spaces around the other operators.

hollyhan commented:
Thanks for the suggestion and the note on rebase, @xylar!

I've updated the lines setting up `cell_width` and also rebased the branch to `origin/master`.

hollyhan commented:
> Looks good to me. Happy to re-test if you like.

Re-testing would be great, thanks so much, @xylar!

hollyhan commented:
> When testing this, I'm seeing the issue I created in MPAS-Tools: [MPAS-Dev/MPAS-Tools#454](https://github.com/MPAS-Dev/MPAS-Tools/issues/454) about the function `pyproj.transform()` being deprecated.

This issue has been resolved by [https://github.com/MPAS-Dev/MPAS-Tools/pull/459](url)

xylar commented:
> This issue has been resolved by [https://github.com/MPAS-Dev/MPAS-Tools/pull/459]

Great! We will need to release MPAS-Tools and update compass to use it before we will notice that here, but that's fine.

matthewhoffman commented on compass/landice/__init__.py:
```diff
@@ -1,4 +1,5 @@
 from compass.mpas_core import MpasCore
+from compass.landice.tests.antarctic import Antarctic
```
I'd prefer we call this new test group `Antarctica` instead of `Antarctic`.  I realize that's kind of an annoying, picky comment because it requires touching a lot of different files, but this could end up being a large and often used test group, so I think it's worth requesting.

matthewhoffman commented on compass/landice/tests/antarctic/default/__init__.py:
```diff
@@ -0,0 +1,28 @@
+from compass.landice.tests.antarctic.mesh import Mesh
+
+
+class Default(TestCase):
```
I think we should rename this to something more descriptive.  Eventually we may have a handful of AIS mesh generation test cases and possibly various tests for running AIS simulations.  How about we rename this to `lowResMesh` or maybe more generically `meshGen` if we think we will support multiple resolutions for the same mesh generation test case (which I think will be the case).

xylar replied:
I'm fine with this.  To be clear, it would need to be `LowResMesh` or `MeshGen` (starting with a capital letter).

xylar replied:
I think `MeshGen` is better because low res is just the default configuration of this test case.

matthewhoffman commented on compass/landice/tests/antarctic/mesh.py:
```diff
@@ -0,0 +1,219 @@
+        dsMesh = convert(dsMesh, logger=logger)
+        logger.info('writing grid_converted.nc')
+        write_netcdf(dsMesh, 'grid_converted.nc')
+        # If no number of levels specified in config file, use 10
```
```suggestion
        # If no number of levels specified in config file, use default
```
(default currently isn't 10)

trhille replied:
Technically, the default is whatever is specified in the config file, so this comment should probably be removed.

matthewhoffman reviewed (changes requested):
@hollyhan , thanks for getting this PR in good shape.  I made a couple suggestions about renaming some of the pieces.  If anyone (you or @trhille or @xylar ) objects, we can discuss.

hollyhan commented:
@matthewhoffmanm, @trhille, @xylar. 
Sorry for the delay in addressing the most recent requested changes. Here is the updated PR - I've tested it and confirmed the mesh generation works as expected.

matthewhoffman commented:
How long does this normally take to run (on Badger)?  It's been running for about 15 minutes, and I don't see any log output beyond:
```
$ compass run
landice/antarctica/mesh_gen
compass calling: compass.landice.tests.antarctica.mesh_gen.MeshGen.run()
  inherited from: compass.testcase.TestCase.run()
  in /turquoise/usr/projects/climate/mhoffman/compass/compass/compass/testcase.py

Running steps: mesh
  * step: mesh
```

In the run directory, it looks like some files are being generated, but just slowly.  These are the files that have appeared since the job started:
```
-rw-rw---- 1 mhoffman climate 5.4M Apr 26 21:33 cellWidthVsXY.nc
-rw-rw---- 1 mhoffman climate  197 Apr 26 21:33 mesh.msh
-rw-rw---- 1 mhoffman climate  236 Apr 26 21:33 mesh.jig
-rw-rw---- 1 mhoffman climate 4.6M Apr 26 21:33 mesh-HFUN.msh
-rw-rw---- 1 mhoffman climate  51M Apr 26 21:35 mesh-MESH.msh
-rw-rw---- 1 mhoffman climate 4.4K Apr 26 21:35 mesh.log
-rw-rw---- 1 mhoffman climate  61M Apr 26 21:36 mesh_triangles.nc
-rw-rw---- 1 mhoffman climate 781M Apr 26 21:37 base_mesh.nc
-rw-rw---- 1 mhoffman climate 781M Apr 26 21:38 grid_converted.nc
-rw-rw---- 1 mhoffman climate  43M Apr 26 21:39 antarctica_8km_2020_10_20_floodFillMask.nc
drwxrwx--- 2 mhoffman climate  13K Apr 26 21:39 .
-rw-rw---- 1 mhoffman climate 808M Apr 26 21:42 ais_8km_preCull.nc
```
Is that normal?  Or is maybe my environment messed up or something?

hollyhan commented:
> How long does this normally take to run (on Badger)? It's been running for about 15 minutes, and I don't see any log output beyond:
> 
> ```
> $ compass run
> landice/antarctica/mesh_gen
> compass calling: compass.landice.tests.antarctica.mesh_gen.MeshGen.run()
>   inherited from: compass.testcase.TestCase.run()
>   in /turquoise/usr/projects/climate/mhoffman/compass/compass/compass/testcase.py
> 
> Running steps: mesh
>   * step: mesh
> ```
> 
> In the run directory, it looks like some files are being generated, but just slowly. These are the files that have appeared since the job started:
> 
> ```
> -rw-rw---- 1 mhoffman climate 5.4M Apr 26 21:33 cellWidthVsXY.nc
> -rw-rw---- 1 mhoffman climate  197 Apr 26 21:33 mesh.msh
> -rw-rw---- 1 mhoffman climate  236 Apr 26 21:33 mesh.jig
> -rw-rw---- 1 mhoffman climate 4.6M Apr 26 21:33 mesh-HFUN.msh
> -rw-rw---- 1 mhoffman climate  51M Apr 26 21:35 mesh-MESH.msh
> -rw-rw---- 1 mhoffman climate 4.4K Apr 26 21:35 mesh.log
> -rw-rw---- 1 mhoffman climate  61M Apr 26 21:36 mesh_triangles.nc
> -rw-rw---- 1 mhoffman climate 781M Apr 26 21:37 base_mesh.nc
> -rw-rw---- 1 mhoffman climate 781M Apr 26 21:38 grid_converted.nc
> -rw-rw---- 1 mhoffman climate  43M Apr 26 21:39 antarctica_8km_2020_10_20_floodFillMask.nc
> drwxrwx--- 2 mhoffman climate  13K Apr 26 21:39 .
> -rw-rw---- 1 mhoffman climate 808M Apr 26 21:42 ais_8km_preCull.nc
> ```
> 
> Is that normal? Or is maybe my environment messed up or something?

@matthewhoffman  Hmm... that's interesting.. It looks like everything seems to work fine and not much slower than my testing (on my laptop) until up to the file `.`.   I don't have such a file generated. Here is my output list:

```
-rw-r--r--  1 hollyhan  staff    5585156 Apr 25 10:48 cellWidthVsXY.nc
-rw-r--r--  1 hollyhan  staff    4767202 Apr 25 10:48 mesh-HFUN.msh
-rw-r--r--  1 hollyhan  staff        197 Apr 25 10:48 mesh.msh
-rw-r--r--  1 hollyhan  staff        236 Apr 25 10:48 mesh.jig
-rw-r--r--  1 hollyhan  staff   52797247 Apr 25 10:50 mesh-MESH.msh
-rw-r--r--  1 hollyhan  staff       4432 Apr 25 10:50 mesh.log
-rw-r--r--  1 hollyhan  staff   63230920 Apr 25 10:51 mesh_triangles.nc
-rw-r--r--  1 hollyhan  staff  818682048 Apr 25 10:52 base_mesh.nc
-rw-r--r--  1 hollyhan  staff  818056204 Apr 25 10:53 grid_converted.nc
-rw-r--r--  1 hollyhan  staff   44625647 Apr 25 10:53 antarctica_8km_2020_10_20_floodFillMask.nc
-rw-r--r--  1 hollyhan  staff  849546212 Apr 25 10:57 ais_8km_preCull.nc
-rw-r--r--  1 hollyhan  staff  103260832 Apr 25 10:57 antarctica_culled.nc
-rw-r--r--  1 hollyhan  staff  108073300 Apr 25 10:57 antarctica_dehorned.nc
-rw-r--r--  1 hollyhan  staff  129637940 Apr 25 10:57 Antarctica_8to80km.nc
-rw-r--r--  1 hollyhan  staff    2836021 Apr 25 10:57 graph.info
```
Rest of the files after (including) `ais_8km_preCull.nc` are generated in less than a minute, Maybe something is going on in Badger that has to do with the `.` file?

xylar commented:
@hollyhan, `.` is just the "file" associated with the current directory.  So that's not likely to be the issue.

matthewhoffman commented on compass/landice/tests/antarctica/mesh.py:
```diff
@@ -0,0 +1,218 @@
+        if float(cullDistance) > 0.:
+            logger.info('calling define_cullMask.py')
+            args = ['define_cullMask.py', '-f',
+                    'ais_8km_preCull.nc', '-m'
+                                          'distance', '-d', cullDistance]
```
```suggestion
                    'ais_8km_preCull.nc', '-m',
                    'distance', '-d', cullDistance]
```
There was a missing comma here - the culling might not be doing what we think it is doing, because the `-m` is probably being ignored: `Running: define_cullMask.py -f ais_8km_preCull.nc -mdistance -d 70.0`
I only noticed this because this step takes so long, and I was staring at it on my machine for a long time. :)

I'll rerun after making this change locally and confirm.

xylar replied:
This is a good reason to switch from the `OptionParser` to the `ArgumentParser` as a general rule.  This would be an error in the `ArgumentParser`.

xylar replied:
That comment is obviously outside the scope of this PR...

hollyhan replied:
Oh! Thanks for spotting that @matthewhoffman! Hmm I wonder how all the previous tests were running fine in other machines. 

This also means that the same line of the `mesh.py` code in humboldt test case wants to be fixed: 
https://github.com/MPAS-Dev/compass/blob/master/compass/landice/tests/humboldt/mesh.py#L105.

I'm happy to make a separate PR fixing this, or simply include in the current PR even though this PR is for Antarctic mesh. What do you think?

matthewhoffman commented:
@hollyhan , I ran this again this morning, and now I'm seeing the full log output in my terminal as it runs.  Last night the log information wasn't appearing.  I have no idea why, but I'm going to assume it was operator error or a temporary problem on Badger.  So I am able to successfully run this to completion with the expected log output and the resulting mesh looks as anticipated.  Nice work.  There are some weird field values in the final file, but that's coming from the source data file in the interpolation, not anything in your workflow.  I'll follow up with you and Trevor on Slack about if/how we want to address that before merging.

hollyhan commented:
Thanks @matthewhoffman. The current head now includes the missing comma and the fix in spacing.

trhille reviewed (changes requested):
Great work on this @hollyhan! I requested a few small changes, but they are all essentially trivial.
My testing passed successfully, and the mesh looks as expected.
```
landice/antarctica/mesh_gen
  * step: mesh
  test execution:      SUCCESS
Test Runtimes:
11:53 PASS landice_antarctica_mesh_gen
Total runtime 11:54
PASS: All passed successfully!
```

hollyhan commented:
>Great work on this @hollyhan! I requested a few small changes, but they are all essentially trivial.
My testing passed successfully, and the mesh looks as expected.

Thanks for the review, @trhille! I've made the changes as requested (3250c042bc7138e9142e1a13e602bfd0a854dd54)

trhille commented on compass/landice/tests/antarctica/mesh.py:
```diff
@@ -115,8 +113,7 @@ def run(self):
 
-        # This step is only necessary because the GeoJSON region
-        # is defined by lat-lon.
+        # Set lat-lon fields in the pre-culled mesh
```
You can compare with the new Greenland mesh creation test case, which also does not use a GeoJSON file: https://github.com/MPAS-Dev/compass/blob/master/compass/landice/tests/greenland/mesh.py#L90-L100

trhille reviewed (changes requested):
@hollyhan, thanks for implementing these changes. I requested one more change that I missed in my review yesterday. 
Also, it looks like your force-push got rid of Matt's commit that added a comma in the Humboldt case. Please bring that commit back into this PR.

xylar commented:
@hollyhan, to get @matthewhoffman's commit back, you can do:
```
git cherry-pick 347b0e1f95543987c7ddd7628314ba96067a7cc2
```
I got the commit hash by looking through my email for when @matthewhoffman pushed that commit.  It can sometimes be tricky to find a "lost" commit.  Making the same changes yourself is okay but getting back the original commit is better.

hollyhan commented:
> @hollyhan, to get @matthewhoffman's commit back, you can do:
> 
> ```
> git cherry-pick 347b0e1f95543987c7ddd7628314ba96067a7cc2
> ```
> 
> I got the commit hash by looking through my email for when @matthewhoffman pushed that commit. It can sometimes be tricky to find a "lost" commit. Making the same changes yourself is okay but getting back the original commit is better.

Thanks a lot for this, Xylar! I brought the commit back by cherry-picking.

trhille reviewed (approved):
Thanks for humoring me, @hollyhan! My testing confirms that all these changes worked as expected. I also tested the Humboldt mesh creation case just to be sure, since one commit here touches that case, and it looks good.

hollyhan commented:
Thanks for testing the PR again and approving it, @trhille! I also made the last changes that we talked about in deleting the inapplicable comment (cc2bd353e66886a81dc70033017443ea36b74e32). 

@xylar, this is all ready to be merged now!

xylar commented:
Great, merging!
